### PR TITLE
shell: Add IPC Service backend with samples

### PR DESCRIPTION
**Pull request by KAGA164**
This PR contains following changes:

- IPC Service backend for shell
- The Remote shell sample for nRF5340
- The configuration for the Radio Test sample that adds possibility to use shell through application core with a USB interface

### CODE REVIEW

**Review by nordic-krch — COMMENTED** [review]
Couple of comments.
I wonder if content of app sample could be extracted to something like `uart_ipc_proxy` module which could be reused.

File: subsys/shell/shell_ipc.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	.enable = enable,
+	.write = write,
+	.read = read,
+};
+
+static int enable_shell_ipc(const struct device *arg)
+{
+	ARG_UNUSED(arg);
+
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ipc0));
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply subsys/shell/shell_ipc.c:144]: Right. I forgot to change this after testing. Will do

File: doc/nrf/libraries/shell/shell_ipc.rst [hunk 2]
```diff
@@ -0,0 +1,23 @@
+.. _shell_ipc_readme:
+
+IPC service shell transport
+###########################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The IPC Service shell transport allows you to receive shell commands through the other remote processor.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply doc/nrf/libraries/shell/shell_ipc.rst]: Done

File: doc/nrf/releases/release-notes-changelog.rst [hunk 3]
```diff
@@ -166,6 +173,15 @@ Gazell samples
...
     There are now four different sample pages, where each Host sample must be used along with its corresponding Device sample.
 
+Other Samples
+-------------
+
+* Updated:
+
+  * :ref:`radio_test`:
+
+    * Added new configuration that builds sample with support for remote IPC Service shell on nRF5340 application core through a USB.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply doc/nrf/releases/release-notes-changelog.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Overview
+********
+
+The sample shows how to run remote shell on the remote cpu using IPC service backend for shell.
+It collects shell data from the IPC endpoint and forwards it to the terminal by the USB transport or the UART.
+If shell is ran on a network core then you might want to use a peripheral that the network core does not have like for example an USB.
+You can use this sample as the reference how to forward shell data from a network core to a application core and connect the shell terminal through application core.
+
+Requirements
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst:12]: Sure, done

File: samples/nrf5340/remote_shell/README.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+.. contents::
+   :local:
+   :depth: 2
+
+You can use this sample to run the remote shell on the nRF5340 network core through the IPC service.
+
+Overview
+********
+
+The sample shows how to run remote shell on the remote cpu using IPC service backend for shell.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+It collects shell data from the IPC endpoint and forwards it to the terminal by the USB transport or the UART.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+If shell is ran on a network core then you might want to use a peripheral that the network core does not have like for example an USB.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+You can use this sample as the reference how to forward shell data from a network core to a application core and connect the shell terminal through application core.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+.. _remote_shell_testing:
+
+Testing
+=======
+
+To test the sample, complete the following steps:
+
+1. Program the sample to the application core.
+#. Program the :ref:`radio_test` sample to the network core.
+#. Plug the DK into an USB host device.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+   The DK is visible as a COM port (Windows) or ttyACM device (Linux) after you connect the development kit over USB
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: samples/nrf5340/remote_shell/README.rst [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+#. |connect_terminal|
+
+Dependencies
+************
+
+It uses the following Zephyr libraries:
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/README.rst]: Done

File: include/shell/shell_ipc.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+	struct shell_transport _name = { \
+	       .api = &shell_ipc_transport_api, \
+	       .ctx = &_name##shell_ipc \
+	}
+
+/**
+ * @brief This function provides pointer to shell IPC Service backend instance.
+ *
+ * Function returns pointer to the shell IPC Service instance. This instance can be
+ * next used with shell_execute_cmd function in order to test commands behavior.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/shell/shell_ipc.h]: Done

File: include/shell/shell_ipc.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+		.rx_ringbuf = &_name##_rx_ring_buffer \
+	}; \
+					    \
+	struct shell_transport _name = { \
+	       .api = &shell_ipc_transport_api, \
+	       .ctx = &_name##shell_ipc \
+	}
+
+/**
+ * @brief This function provides pointer to shell IPC Service backend instance.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/shell/shell_ipc.h]: Done

File: doc/nrf/libraries/shell/shell_ipc.rst [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+It uses the IPC Service feature.
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply doc/nrf/libraries/shell/shell_ipc.rst:11]: I didn't find any doc for the IPC Service which I could link here. The `zephyr:IPC_Service_sample` uses the IPC Service similar like we do here

File: samples/nrf5340/remote_shell/src/main.c [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+	uart_irq_tx_enable(uart_dev);
+}
+
+static void uart_irq_handler(const struct device *dev, void *user_data)
+{
+	int err;
+	const struct uart_irq_context *uart_ctrl = user_data;
+
+	uart_irq_update(dev);
+
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply samples/nrf5340/remote_shell/src/main.c:86]: Done

File: subsys/shell/shell_ipc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (KAGA164) [reply subsys/shell/shell_ipc.c]: Done

File: include/shell/shell_ipc.h [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+ * @defgroup shell_ipc IPC service shell transport
+ * @{
+ * @brief IPC Service shell transport API.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern const struct shell_transport_api shell_ipc_transport_api;
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply include/shell/shell_ipc.h]: Ok I removed all macros that could be used to create an instance and moved all instance data structures into source file as discussed offline. For architectures with more than 2 CPUs we can provide a different solution directly in the shell subsystem but this backend still can be used for remote CPUs but I don't see any use case where more than one IPC backend on CPU could be used.

**Comment by DatGizmo** [comment]
With the new Integration Test Specification process we don't need the CI-* github labels.
I will remove them.

To get the test spec feature working correctly on this PR, you need to rebase it to the latest main to get the [.github/test-spec.yml](https://github.com/nrfconnect/sdk-nrf/blob/main/.github/test-spec.yml) file

Then run CI here to update the test spec.

[Information about the test spec feature](https://projecttools.nordicsemi.no/confluence/display/NCS/Test+Spec)
[CI-Help for issues](https://ci-help.nordicsemi.no/)

**Review by nordic-krch — APPROVED** [review]

File: include/shell/shell_ipc.h [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef SHELL_IPC_H_
+#define SHELL_IPC_H_
+
+#include <kernel.h>
+#include <ipc/ipc_service.h>
+#include <shell/shell.h>
+#include <sys/ring_buffer.h>
```
Comment (kapi-no) [inline include/shell/shell_ipc.h:13]: Limit the number of unnecessary includes. If you need them in the implementation unit then add them there

```suggestion
#include <shell/shell.h>
```

File: samples/nrf5340/remote_shell/src/main.c [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+		return;
+	}
+
+	if (IS_ENABLED(CONFIG_USB_DEVICE_STACK)) {
+		err = usb_enable(NULL);
+		if (err != 0) {
+			LOG_ERR("Failed to enable USB, err %d", err);
+			return;
+		}
+	}
```
Comment (kapi-no) [inline samples/nrf5340/remote_shell/src/main.c:149]: Why do you enable USB module here? Is it used by the application?

**Review by kapi-no — COMMENTED** [review]
Two nits

**Review by kapi-no — APPROVED** [review]

**Pull request merged**